internal/cedana/streamer/dump_adapters.go

```diff
@@ -116,13 +135,96 @@ func DumpFilesystem(streams int32) types.Adapter[types.Dump] {
 
 			// XXX: We do not differentiate between leave-running or not, because unfortunately CRIU
 			// does not close the streaming file descriptors on its side when the PostDumpFunc is triggered.
-			// This is why the logic here is not the same as that in `filesystem/dump_adapters.go`.
+			// This is why the logic here is not the same as that in `filesystem/dump_adapters.go`
+			if async {
+				ext, _ := cedana_io.ExtForCompression(compression)
+
+				upload := func(ctx context.Context) error {
+					var wg sync.WaitGroup
+					errCh := make(chan error, streams)
+
+					for i := int32(0); i < streams; i++ {
+						wg.Add(1)
+						go func(i int32) {
+							defer wg.Done()
+
+							localPath := fmt.Sprintf("%s/img-%d%s", imagesDirectory, i, ext)
+							remotePath := fmt.Sprintf("%s/img-%d%s", path, i, ext)
+
+							src, err := streamStorage.Open(ctx, localPath)
+							if err != nil {
+								errCh <- fmt.Errorf("failed to open local shard %d: %w", i, err)
+								return
+							}
+							defer src.Close()
+
+							dst, err := storage.Create(ctx, remotePath)
+							if err != nil {
+								errCh <- fmt.Errorf("failed to create remote shard %d: %w", i, err)
+								return
+							}
+							defer dst.Close()
+
+							if _, err := io.Copy(dst, src); err != nil {
+								errCh <- fmt.Errorf("failed to upload shard %d: %w", i, err)
+							}
+						}(i)
+					}
+
+					wg.Wait()
+					close(errCh)
+
+					var uploadErr error
+					for e := range errCh {
+						uploadErr = errors.Join(uploadErr, e)
+					}
 
-			defer func() {
-				_, end := profiling.StartTimingCategory(ctx, "storage", waitForIO)
-				err = errors.Join(err, waitForIO())
-				end()
-			}()
+					os.RemoveAll(imagesDirectory)
+					return uploadErr
+				}
+
+				defer func() {
+					_, end := profiling.StartTimingCategory(ctx, "storage", waitForIO)
+					err = errors.Join(err, waitForIO())
+					end()
```

Comment:
NIT: This is being called in both branches of this if condition. Let’s simply add this before line 139 (without the defer), and simplify.